File: mcspring-api/mcspring-subcommands/src/main/java/in/kyle/mcspring/subcommands/tab/TabDiscovery.java
package in.kyle.mcspring.subcommands.tab;

//@Component
//@RequiredArgsConstructor
//public class TabDiscovery {
//
//    private final SimpleMethodInjection injection;
//
//    public List<String> getCompletions(CommandSender sender,
//                                       String commandString,
//                                       Consumer<PluginCommand> consumer) {
//        if (!commandString.isEmpty() && !commandString.endsWith(" ")) {
//            // TODO: 2020-02-25 Enable partial completions
//            return Collections.emptyList();
//        }
//        var parts = new ArrayList<>(Arrays.asList(commandString.split(" ")));
//        if (parts.get(0).isEmpty()) {
//            parts.remove(0);
//        }
//        var command = new PluginCommandTabCompletable(injection, sender, parts);
//        consumer.accept(command);
//
//        return getCompletions(command);
//    }
//
//    private List<String> getCompletions(PluginCommandTabCompletable command) {
//        if (command.hasChild()) {
//            return getCompletions(command.getChild());
//        } else if (command.getState() == State.MISSING_ARG || command.getState() == State.CLEAN) {
//            return command.getTabCompletionOptions();
//        } else {
//            return Collections.emptyList();
//        }
//    }
//}

File: mcspring-api/mcspring-subcommands/src/main/java/in/kyle/mcspring/subcommands/tab/TabDiscovery.kt
package `in`.kyle.mcspring.subcommands.tab

import `in`.kyle.mcspring.command.SimpleMethodInjection
import `in`.kyle.mcspring.subcommands.plugincommand.impl.PluginCommandBase.State
import `in`.kyle.mcspring.subcommands.plugincommand.impl.PluginCommandImpl
import `in`.kyle.mcspring.subcommands.plugincommand.api.PluginCommand
import org.bukkit.command.CommandSender
import org.springframework.stereotype.Component

@Component
class TabDiscovery(
        private val injection: SimpleMethodInjection
) {
    fun getCompletions(sender: CommandSender,
                       commandString: String,
                       consumer: (PluginCommand) -> Unit): List<String> {
        val (parts, prefix) = getCommandParts(commandString)

        val command = PluginCommandImpl(injection, sender, parts, false)
        consumer(command)

        val completionCandidates = getCompletions(command)
        return completionCandidates.filter { it.startsWith(prefix) }
    }

    private fun getCommandParts(command: String): Pair<MutableList<String>, String> {
        val parts = command.split(" ").filter { it.isNotBlank() }.toMutableList()

        val prefix = if (!command.endsWith(" ") && command.isNotBlank()) {
            parts.removeAt(parts.size - 1)
        } else {
            ""
        }
        return Pair(parts, prefix)
    }

    private fun getCompletions(command: PluginCommandImpl): List<String> {
        return if (command.child != null) {
            getCompletions(command.child as PluginCommandImpl)
        } else if (command.state == State.MISSING_ARG || command.state == State.CLEAN) {
            command.completions.flatMap { it.completions }
        } else {
            emptyList()
        }
    }
}
